isisd/isis_spf_delay.c

```diff
@@ -0,0 +1,457 @@
+/*
+ * This is an implementation of the IETF SPF delay algorithm 
+ * as explain in draft-ietf-rtgwg-backoff-algo
+ *
+ * Module name: isis_spf_delay.c
+ * Version:     0.1
+ * Created:     25-01-2017 by S. Litkowski 
+ * Copyright (C) 2017 Orange Labs http://www.orange.com/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
+ * MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+#include <sys/time.h>
+#include <stdlib.h>
+
+#include "thread.h"
+
+#include "prefix.h"
+
+#include "isisd/isis_constants.h"
+#include "isisd/isisd.h"
+#include "isisd/isis_adjacency.h"
+#include "isisd/isis_spf.h"
+#include "isisd/isis_spf_delay.h"
+
+const char       *isis_spf_delay_states[3] = { "QUIET", 
+					       "SHORT_WAIT", 
+					       "LONG_WAIT"};
+
+char *
+str_mtimestr(time_t uptime)
+{
+
+#define ONE_MINUTE_SECOND 60*60
+#define ONE_DAY_SECOND 60*60*24
+#define ONE_WEEK_SECOND 60*60*24*7
+
+  struct tm *tm;
+  char *buf = (char *)malloc(15);
+
+  tm = gmtime (&uptime);
+  strftime(buf, 15, "%Z %H:%M:%S", tm);
+
+  return buf;
+}
+
+
+int
+isis_spf_delay_event_ttl_expires(struct thread *thread) {
+
+  int level = 0;
+  struct isis_spf_delay_reinit    *reinit;
+  struct isis_area        *area;
+  int family = 0;
+
+  reinit = THREAD_ARG(thread);
+  assert(reinit);
```

Comment:
I keep seeing this type of assert and I really don't get the point.  1) thread->arg if we are calling this function cannot be NULL 2) If it is NULL something has gone so terribly wrong you are crashing no matter what.